UC480 Thorlabs camera driver exposure facets, thorlabs TLPM powermeter, smaract SCU actuator drivers

seb5g opened the pull request.

seb5g commented:
Hello,
I pushed some updates on smaract and some patches on the K10CR1 driver from thorlabs, could you merge them. Also, what's your next release planned?

Also I tried to create a driver on thorlabs new powermeter library TLPM but I'm having issues in parsing the header file. I ended writing my own driver using ctypes but it's kind of a waste... Could you provide some help?

natezb commented:
Sorry for the delay, things have been busy. I'll try to look over this soon. Are you still having difficulties with the TLPM header file?

seb5g commented:
Concerning the TLPM, I couldn't manage to make it work. I used classical ctypes way to create a pymodaq's plugin. Wish I could add this driver to instrumental but I'm kind of stuck...

natezb commented:
Do you have an error traceback or something that you could post?

seb5g commented:
Here is the traceback:
```
Connected to pydev debugger (build 192.7142.42)
C:\Miniconda3\envs\pymodaq_dev\lib\site-packages\nicelib\process.py:1472: UserWarning: Un-pythonable macro _VI_PTR: <generator>:2:30: before: ;
  warnings.warn("Un-pythonable macro {}: {}".format(macro.name, str(e)))
C:\Miniconda3\envs\pymodaq_dev\lib\site-packages\nicelib\process.py:1472: UserWarning: Un-pythonable macro ViPtr: <generator>:2:28: before: ;
  warnings.warn("Un-pythonable macro {}: {}".format(macro.name, str(e)))
Traceback (most recent call last):
  File "C:\Users\weber\AppData\Roaming\Python\Python38\site-packages\cffi\api.py", line 126, in _cdef
    self._parser.parse(csource, override=override, **options)
  File "C:\Users\weber\AppData\Roaming\Python\Python38\site-packages\cffi\cparser.py", line 389, in parse
    self._internal_parse(csource)
  File "C:\Users\weber\AppData\Roaming\Python\Python38\site-packages\cffi\cparser.py", line 412, in _internal_parse
    self._parse_decl(decl)
  File "C:\Users\weber\AppData\Roaming\Python\Python38\site-packages\cffi\cparser.py", line 505, in _parse_decl
    self._declare_function(tp, quals, decl)
  File "C:\Users\weber\AppData\Roaming\Python\Python38\site-packages\cffi\cparser.py", line 498, in _declare_function
    self._declare(tag + decl.name, tp)
  File "C:\Users\weber\AppData\Roaming\Python\Python38\site-packages\cffi\cparser.py", line 568, in _declare
    raise FFIError(
cffi.FFIError: multiple declarations of function TLPM_setDigIoOutput (for interactive usage, try cdef(xx, override=True))

Process finished with exit code 1
```

and this is the code:

```
from nicelib import build_lib, generate_bindings
from nicelib.process import declspec_hook, struct_func_hook
import os


header_info = {
    'win32': {
        'path': (
            r"{VXIPNPPATH}\WinNT\Include",
        ),
        'header': 'TLPM.h'
    },
    'win64': {
        'path': (
            r"{VXIPNPPATH64}\Win64\Include",
        ),
        'header': 'TLPM.h'
    },
}

lib_names = {'win32': f'{os.environ["VXIPNPPATH"]}\\WinNT\\Bin\\TLPM_32.dll',
             'win64': f'{os.environ["VXIPNPPATH64"]}\\Win64\\Bin\\TLPM_64.dll'}


def fastcall_hook(tokens):
    """Removes ``cdecl``, ``_cdecl``, and ``__cdecl``

    Enabled by default.
    """
    for token in tokens:
        if token not in ('__fastcall',):
            yield token

def build():
    build_lib(header_info, lib_names, '_tlpmlib', __file__, token_hooks=(declspec_hook, fastcall_hook))


# def bindings():
#     with open('bindings.py') as f:
#         generate_bindings(header_info, f)


if __name__ == '__main__':
    build()
    #bindings()
```

seb5g commented:
and the header file:

[TLPM.txt](https://github.com/mabuchilab/Instrumental/files/7994477/TLPM.txt)

natezb commented:
The header you're trying to parse has a repeated declaration, and CFFI doesn't like that. I just pushed a change to NiceLib (mabuchilab/NiceLib@feb91e9001a56619f8d4a956743c8df3ceb6a2d2) that adds an `override` option to `build_lib()`. In  your case you should try passing `override=True`, which should allow the repeated declaration.

seb5g commented:
Hello there, using your new commit, I manage to build from the header and produced this new driver. Compatible with the new thorlabs brand of powermeters and tested with the PM101A. 

I also produced a new driver for the motion stages from mad city labs

seb5g commented:
by the way, I'm organising in next October a session of 5 days around my PyMoDAQ framework (pymodaq.cnrs.fr). 3 days will be devoted to training of new users while the remaining two days will be a kind of PyMoDAQ Days. With presentations of people using PyMoDAQ in their lab, round table about issues and new developement. I would also like to have an talk from people who are developing libraries such as Instrumental. It would be nice to have some input from people who are on the field of open-source software for lab instrumentation for some time and also see if/how bridges could be build between instrumental and pymodaq. Would you be interested?

seb5g commented:
Hello,
I'm sure you are quite busy, but I'm making much effort in adding new driver to the library, but I'm now facing the problem that I cannot publish pypi updates of my instrument plugin using instrumental new features because there is no new release of instrumental and couldn't find any schedule.

Did you have thought into my invitation?

natezb commented on instrumental/drivers/motion/_kinesis/isc.py:
```diff
@@ -73,22 +73,44 @@ def _start_polling(self, polling_period='200ms'):
         self.polling_period = polling_period
         self.dev.StartPolling(self.polling_period.m_as('ms'))
 
+    def get_info(self):
```
This seems unnecessary. Can you use something like `ffi.string()` instead of adding this new `decode` method?

natezb commented on instrumental/drivers/motion/_madcitylabs/nanodrive.py:
```diff
@@ -0,0 +1,144 @@
+
+log = get_logger(__name__)
+
+_INST_PARAMS = ['serial']
+_INST_CLASSES = ['NanoDrive']
```
I'm not sure if it's documented very well, but at some point we moved to `_INST_*_` defined as class attributes rather than at the module level. For instance, see the [Tektronix driver](https://github.com/mabuchilab/Instrumental/blob/8cfc5a1f8c93a00a5b23881b170b3fb191bb2087/instrumental/drivers/scopes/tektronix.py).

seb5g replied:
Guess That answer my question above, You should modify the docs concerning this point. Should we still use _INST_CLASSES then?

natezb replied:
No, you shouldn't need `_INST_CLASSES` anymore. The parser automatically includes classes that define other `_INST_*_` class attributes.

seb5g replied:
got it and modified the doc accordingly

natezb commented on instrumental/drivers/motion/_smaract/lextab.py:
```diff
@@ -0,0 +1,10 @@
+# lextab.py. This file automatically created by PLY (version 3.10). Don't edit!
+_tabversion   = '3.10'
+_lextokens    = set(('SIGNED', 'FLOAT_CONST', 'SEMI', 'PERIOD', 'SHORT', 'CONST', 'UNSIGNED', 'CHAR_CONST', 'INT_CONST_HEX', 'RSHIFTEQUAL', 'LE', 'CONTINUE', 'EXTERN', 'WSTRING_LITERAL', 'VOLATILE', 'MINUS', 'PLUS', 'MINUSEQUAL', 'ENUM', 'WHILE', '_COMPLEX', 'ARROW', 'INT_CONST_CHAR', 'INT_CONST_BIN', 'LNOT', 'CHAR', 'LAND', 'OREQUAL', 'MOD', 'TYPEID', 'RBRACE', 'ID', 'LPAREN', 'LT', 'TIMES', 'DE ...
+_lexreflags   = 64
+_lexliterals  = ''
+_lexstateinfo = {'INITIAL': 'inclusive', 'ppline': 'exclusive', 'pppragma': 'exclusive'}
+_lexstatere   = {'INITIAL': [('(?P<t_PPHASH>[ \\t]*\\#)|(?P<t_NEWLINE>\\n+)|(?P<t_LBRACE>\\{)|(?P<t_RBRACE>\\})|(?P<t_FLOAT_CONST>((((([0-9]*\\.[0-9]+)|([0-9]+\\.))([eE][-+]?[0-9]+)?)|([0-9]+([eE][-+]?[0-9]+)))[FfLl]?))|(?P<t_HEX_FLOAT_CONST>(0[xX]([0-9a-fA-F]+|((([0-9a-fA-F]+)?\\.[0-9a-fA-F]+)|([0-9a-fA-F]+\\.)))([pP][+-]?[0-9]+)[FfLl]?))|(?P<t_INT_CONST_HEX>0[xX][0-9a-fA-F]+(([uU]ll)|([uU]LL)|( ...
+_lexstateignore = {'INITIAL': ' \t', 'ppline': ' \t', 'pppragma': ' \t'}
+_lexstateerrorf = {'INITIAL': 't_error', 'ppline': 't_ppline_error', 'pppragma': 't_pppragma_error'}
+_lexstateeoff = {}
```
This file should be removed

seb5g replied:
done

natezb commented on instrumental/drivers/motion/_smaract/yacctab.py:
```diff
@@ -0,0 +1,343 @@
+  ('brace_open -> LBRACE','brace_open',1,'p_brace_open','c_parser.py',1838),
+  ('brace_close -> RBRACE','brace_close',1,'p_brace_close','c_parser.py',1844),
+  ('empty -> <empty>','empty',0,'p_empty','c_parser.py',1850),
+]
```
This file should be removed

seb5g replied:
done

natezb reviewed (changes requested):
I left some comments on the parts for you to take a look at. I only skimmed the new drivers you added, but they should be ok to merge.

natezb commented:
I am sorry for how long it's taken to review this. I haven't been able to do this as part of my day job for several years, so I need to find spare time to address whatever issues there are. I also haven't used the library for a few years and don't have access to the vast majority of equipment, which makes it pretty challenging to verify that things work.

I made a few code review notes, and when those things are addressed I think everything should be good to merge. Sorry again for all the delays, and thanks for your contributions!

As for your proposal, it is interesting but as you can tell I don't have much energy to devote to this project at the moment. I'll think it over.

seb5g commented on .gitignore:
```diff
@@ -20,3 +21,9 @@ Release/
+*yacctab.py
+
+
+instrumental/drivers/motion/_smaract/lextab.py
```
no you're correct, I'll remove it

seb5g commented on instrumental/driver_info.py:
```diff
@@ -130,6 +141,11 @@
+        'params': ['name', 'reset'],
+        'classes': ['TLPM'],
+        'imports': [],
+    }),
```
Ok I didn't know about this parse_modules  functionality (not seen in the doc, as far as I know). Could you just explained to me how the variable _INST_PARAMS and _INST_CLASSES are working together? Could not figure if I should place them in the module level (after imports) or in each classes
![image](https://user-images.githubusercontent.com/6609955/163125525-76fc9aaa-174b-4407-9a97-1cb0de3d4239.png)

natezb replied:
I just looked at the code (`instrumental/drivers/__init__.py`) to jog my memory. The "new" (seemingly undocumented) way is to define `_INST_PARAMS_` (note the _trailing_ underscore) as a class attribute. Then you can remove `_INST_PARAMS` and `_INST_CLASSES` from your module.

I think I mentioned this in another comment, but you can check out the Tektronix oscilloscope driver for an example of how this works.

seb5g replied:
ok I got it, I modified the documentation according to this

seb5g replied:
as for the _INST_PRIORITY priority one, I think it could also be related to one of my modifications, see 71371141b476c442d2317c7e3cd13868335c150b . If one choose one Paramset (with a particular classname in it) then one should be sure this is the one that is instantiated. That was not the case before 71371141b476c442d2317c7e3cd13868335c150b

seb5g commented on instrumental/drivers/motion/_kinesis/common.py:
```diff
@@ -21,8 +21,10 @@ class MessageType(Enum):
+    SettingsExtern = 2
+    Error = 3
+    Close = 4
+    SettingsReset = 5
```
Got this from the docs from a recent installation of kinesis, hence my modifications of the code:
![image](https://user-images.githubusercontent.com/6609955/163125803-8d3ba3af-f9c0-48b2-8d78-b309fc033f6a.png)

seb5g commented:
I addressed most of your points with some questions remaining. Do you know when you will publish a new release on pypi? Do you need some help in maintaining instrumental?

natezb commented:
I can probably make a release on PyPI within the next week. It's been awhile since a release, so some release procedures may have changed--that's the experience I remember from when I released NiceLib a few months back.

And yes, help with maintaining Instrumental would be valuable. There may be some things we could set up like adding flake8 and other checks to automatically run on PRs. Also adding some contribution guidelines.

seb5g commented:
Following your advice for linting on PR I worked on setting up github actions (see commits) but then I just saw you're doing linting on appveyor. While I'm using github actions for my project I added the possibility here. It's very easy to add python versions, and OS as well (ubuntu or windows for instance)
I also added an action (workflow) to trigger automated pypi publication when creating a release on github. It save the hassle to do it manually with twine...

Looking at appveyor config. I saw you dropped python 2.7 and 3.5. But then when I ran flake8 on the code, it found many issues related to python 2 things... I commented some and corrected other but maybe we should just remove them entirely?

Let me know how you would like to work on this?

natezb commented:
Ok, I'll take a look over these further changes. In the future, it's probably better to create a new PR for unrelated changes. Every time more stuff gets added to this PR it takes longer to review and merge it.

I think I previously had automatic uploads to PyPI set up through TravisCI but it may not be working anymore.

My intention was to drop Python 2 support for future releases. But prior releases all had support for both 2 and 3, so there's going to be a lot of code written that way. I think it makes sense for all _new_ code to be targeted at Python 3.6+ and to refactor Py2 stuff away when it makes sense, but bulk changes introduce a lot of unnecessary effort and risk of breakage.

natezb commented on instrumental/drivers/cameras/pco.py:
```diff
@@ -25,8 +25,10 @@
 log = get_logger(__name__)
 
+
 if PY2:
-    memoryview = buffer  # Needed b/c np.frombuffer is broken on memoryviews in PY2
+    # could not find an equivalent in the future module, so I don't knwow what that buffer is??
+    # memoryview = buffer  # Needed b/c np.frombuffer is broken on memoryviews in PY2
     from time import clock
```
I would just leave this as-is. `buffer` is a Python 2 builtin.

seb5g replied:
done

natezb commented on instrumental/drivers/cameras/picam.py:
```diff
@@ -18,8 +18,9 @@
 
-if PY2:
-    memoryview = buffer  # Needed b/c np.frombuffer is broken on memoryviews in PY2
+# could not find an equivalent in the future module, so I don't knwow what that buffer is??
+# if PY2:
+#     memoryview = buffer  # Needed b/c np.frombuffer is broken on memoryviews in PY2
```
Ditto above

seb5g replied:
done

natezb commented on instrumental/drivers/cameras/pixelfly.py:
```diff
@@ -19,7 +19,8 @@
 
 if PY2:
-    memoryview = buffer  # Needed b/c np.frombuffer is broken on memoryviews in PY2
+    # could not find an equivalent in the future module, so I don't knwow what that buffer is??
+    #memoryview = buffer  # Needed b/c np.frombuffer is broken on memoryviews in PY2
```
Ditto above

seb5g replied:
done

natezb commented on instrumental/drivers/cameras/pvcam.py:
```diff
@@ -19,6 +19,10 @@
 from ..util import check_units
 from ...errors import InstrumentTypeError, InstrumentNotFoundError
 from ... import Q_, u
+from future.utils import PY2
+
+
+
```
This doesn't seem needed

seb5g replied:
done

natezb commented on instrumental/drivers/spectrometers/thorlabs_ccs.py:
```diff
@@ -21,7 +21,9 @@
 
 if PY2:
-    memoryview = buffer  # Needed b/c np.frombuffer is broken on memoryviews in PY2
+    # could not find an equivalent in the future module, so I don't knwow what that buffer is??
+    #memoryview = buffer  # Needed b/c np.frombuffer is broken on memoryviews in PY2
+    pass
```
Ditto above

seb5g replied:
done

natezb commented on instrumental/gui.py:
```diff
@@ -354,7 +354,7 @@ def _lut_scale_image(self, arr):
         A = 2**15
         B = 100  # k's scaling factor
-        g = lambda i, k: A*((k-B)*i) / ((2*k)*x - (k+B)*A)
+        #g = lambda i, k: A*((k-B)*i) / ((2*k)*x - (k+B)*A)
```
I want to avoid unnecessary pep8-only changes, so let's revert this.

seb5g replied:
well here, x is just not defined, I tought maybe it should be a parameter of the lambda function. And this method doesn't return anything?

Anyway I reverted it

natezb commented on instrumental/parse_modules.py:
```diff
@@ -302,12 +302,12 @@ def get_imports(source, root):
-        if fname.endswith('.py') and not fname.startswith('_'):
-            mod_name = fname[:-3]
-            yield group + '.' + mod_name
+# def parse_driver_modules(module_name):
+#     driver_dir = os.path.join(THIS_DIR, 'drivers', *(module_name.split('.')))
+#     for fname in os.listdir(driver_dir):
+#         if fname.endswith('.py') and not fname.startswith('_'):
+#             mod_name = fname[:-3]
+#             yield group + '.' + mod_name
```
Ditto pep8 comment above

natezb commented on instrumental/tools.py:
```diff
@@ -461,48 +464,49 @@ def find_FSR():
+#         base_dir = conf.prefs['data_directory']
+#     set_dir = os.path.join(base_dir, date.today().isoformat(), set_name)
+#     if not os.path.exists(set_dir):
+#         os.makedirs(set_dir)
+#
+#     # Block until light is turned on
+#     raw_input('Please turn on light then press [ENTER]: ')
+#
+#     from .drivers.cameras.uc480 import get_camera
+#     top_cam = get_camera(serial=TOP_CAM_SERIAL)
+#     side_cam = get_camera(serial=SIDE_CAM_SERIAL)
+#     top_cam.open()
+#     # top_cam.load_stored_parameters(1)
+#     top_cam.load_stored_parameters(1)
+#     top_cam.save_frame(os.path.join(set_dir, 'Top.jpg'))
+#     top_cam.close()
+#     side_cam.open()
+#     side_cam.load_stored_parameters(1)
+#     side_cam.save_frame(os.path.join(set_dir, 'Side.jpg'))
+#     side_cam.close()
+#
+#     scope = scopes.scope(scopes.SCOPE_A)
+#     fname = 'Ringdown {:02}.csv'
+#     trace_num = 0
+#     cum_FWHM = 0 * u.MHz
+#     print("-------------Enter d[one] to stop taking data-------------")
+#     while True:
+#         s = raw_input('Press [ENTER] to process ringdown {}: '.format(trace_num))
+#         if s and s[0].lower() == 'd':
+#             break
+#         x, y = scope.get_data(channel=1)
+#         full_filename = os.path.join(set_dir, fname.format(trace_num))
+#         _save_data(x, y, full_filename)
+#
+#         FWHM = guided_ringdown_fit(x, y)
+#         _save_summary(full_filename, FWHM)
+#         print("-------------------------------------- FWHM = {}".format(FWHM))
+#         cum_FWHM += FWHM
+#         trace_num += 1
+#     print('Mean FWHM: {}'.format(cum_FWHM/trace_num))
```
Ditto pep8 comment above

seb5g replied:
done

natezb commented on instrumental/tools.py:
```diff
@@ -12,6 +12,9 @@
 from . import u, Q_, conf, instrument
 from .drivers import scopes
 
+
+from past.builtins import basestring, raw_input
+
```
This part seems fine to keep

natezb commented on instrumental/drivers/motion/_kinesis/isc.py:
```diff
@@ -73,22 +75,34 @@ def _start_polling(self, polling_period='200ms'):
 
+    def get_info(self):
+        description = NiceISC.GetDeviceInfo(self.serial).description
+        return ffi.string(description)
+
     @check_units(angle='deg')
-    def move_to(self, angle, wait=False):
-        """Rotate the stage to the given angle
+    def move_to(self, angle, wait=False, move_type='abs'):
+        """Rotate the stage to the given angle in absolute or relative units
 
         Parameters
         ----------
         angle : Quantity
             Angle that the stage will rotate to. Takes the stage offset into account.
+        wait: (bool) if True waits to receive message telling the move is done else returns
+        move_type: (str) either 'abs' (default) for absolute positioning or 'rel' for relative positioning
         """
         log.debug("Moving stage to {}".format(angle))
         log.debug("Current position is {}".format(self.position))
         self.dev.ClearMessageQueue()
-        self.dev.MoveToPosition(self._to_dev_units(angle + self.offset))
+        if move_type == 'abs':
+            self.dev.MoveToPosition(self._to_dev_units(angle + self.offset))
+        else:
+            self.dev.MoveRelative(self._to_dev_units(angle))
```
I thought I posted this before, but maybe I accidentally deleted it. Can you instead create a new `move_relative` method (or perhaps `move_rel` if you prefer)?

seb5g replied:
ok created the move_relative method. On that subject, I would like to create  a common motion interface in the motion object. If you are ok, I would like to introduce this move_to and move_relative in it? I'll make a separated PR for this anyway

natezb reviewed (changes requested):
Ok, I've commented on what should change. Mostly I don't want to start worrying about running flake8 on the entire codebase right now.

Once these changes are done, I think it should be ready to merge.

seb5g commented:
Sorry about the PR mixing everything, it seems my git wasn't configured correctly and pushed everything always on your repo...
